util/svdgen.py

```diff
@@ -0,0 +1,604 @@
+
+    # SVD files require the magic string "%s" in the register to generate
+    # a numbered offset into the window. HJSON doesn't have this; append
+    # it as necessary.
+    name = window.find('./name')
+    if name.text.find('%s') < 0:
+        name.text += '%s'
+
+    return window
+
+def multireg(multi: [hjson]) -> [ET.Element]:
+    """Generate a series of <register> nodes from a multireg. If the
+    multireg is empty this yields no items; if it contains a single item
+    then it yields a single <register> node. When it contains multiple
+    items then a <cluster> element is created containing a <register>
+    node for each."""
+
+    genregs = multi['genregs']
+    if len(genregs) == 1:
+        yield register(genregs[0])
+    elif len(genregs) > 1:
+        # A cluster has a base address, and then registers within the
+        # cluster have address offsets relative to the cluster base.
+        base = min(reg['genoffset'] for reg in genregs)
+
+        yield create('cluster',
+                name          = multi['name'],
+                description   = multi['desc'],
+                addressOffset = hex(base),
+                *(register(reg, base) for reg in genregs))
+
+def registers(regs: [hjson]) -> [ET.Element]:
+    """Convert a list of register HJSON element into a <registers> node.
+    Like generating C headers, this needs to filter based on a register's
+    type.
+
+    This is a generator function; due to filtering reserved/skipped
+    memory and expanding multi-registers the number of registers produced
+    likely will not match the number of registers in."""
+
+    for reg in regs:
+        if 'reserved' in reg or 'skipto' in reg:
+            pass
+        elif 'window' in reg:
+            yield window(reg['window'])
+        elif 'sameaddr' in reg:
+            yield from map(register, reg['sameaddr'])
+        elif 'multireg' in reg:
+            yield from multireg(reg['multireg'])
+        else:
+            yield register(reg)
+
+def peripheral(module: hjson, ip: hjson) -> ET.Element:
+    """Convert an IP module definition to a <peripheral> element. This
+    pulls only the name and base address from the top-level HJSON."""
+
+    # argument ordering is deceptive; <name> and <baseAddress> will
+    # preceed all <interrupt>s, <register>s, and the comment
+    return create('peripheral',
+            *interrupts(ip.get('interrupt_list')),
+            create('registers', *registers(ip['registers'])),
+            ET.Comment('end of %s' % module['name']),
+            name        = module['name'],
+            baseAddress = module['base_addr'])
+
+def cpu() -> ET.Element:
+    """Generate a <cpu> element. Currently all values are hardcoded as
+    they are not included in either the top level or IP module HJSON."""
+
+    # TODO: allow customizing these.
+
+    # RISCV is not present in pysvd so this would fail validation. We
+    # set it to the (inaccurate) value `other`, and patch it up after
```

Comment:
I think it's unfortunately accurate until CMSIS-SVD gets an update that includes RISC-V.